File: agent/node/mapping.ts
import { readFileSync } from "node:fs";

export type Role = "management" | "room" | "reports";
export type Direction = "in" | "out" | "both";
export type Platform = "slack" | "discord";

export type Channel = {
  id: string;
  name: string;
  role: Role;
  direction: Direction;
};

export type Work = {
  id: string;
  name: string;
  room: string;
  state: string;
  anchor: string;
  owner: string;
  pageId?: string;
};

export type Mapping = {
  workspace: string;
  platform: Platform;
  scope: string;
  agent: string;
  collectSeconds: number;
  channels: Channel[];
  work: Work[];
};

export class MappingError extends Error {}

export async function readSource(source: string): Promise<string> {
  if (source.startsWith("http://") || source.startsWith("https://")) {
    const res = await fetch(source, { headers: { "User-Agent": "axiom-relay/0.3" } });
    if (!res.ok) throw new MappingError(`mapping fetch ${res.status}`);
    return await res.text();
  }
  return readFileSync(source, "utf8");
}

export async function loadMappingFrom(source: string): Promise<Mapping> {
  return loadMapping(await readSource(source));
}

export function loadMapping(text: string): Mapping {
  const { meta, body } = splitFrontMatter(text);
  const { channels, work } = parseTables(body);
  validate(meta, channels, work);
  return {
    workspace: meta.workspace,
    platform: meta.platform as Platform,
    scope: meta.scope,
    agent: meta.agent,
    collectSeconds: Number(meta.collect_seconds ?? 45),
    channels,
    work,
  };
}

function splitFrontMatter(text: string): { meta: Record<string, string>; body: string } {
  if (!text.startsWith("---")) throw new MappingError("mapping must start with YAML front matter");
  const end = text.indexOf("\n---", 3);
  if (end < 0) throw new MappingError("unterminated front matter");
  const raw = text.slice(3, end);
  const body = text.slice(end + 4);
  const meta: Record<string, string> = {};
  for (const line of raw.split("\n")) {
    if (!line.trim() || !line.includes(":")) continue;
    const i = line.indexOf(":");
    meta[line.slice(0, i).trim()] = line.slice(i + 1).trim();
  }
  return { meta, body };
}

function parseTables(body: string): { channels: Channel[]; work: Work[] } {
  const channels: Channel[] = [];
  const work: Work[] = [];
  let header: string[] | null = null;
  let rows: Record<string, string>[] = [];

  const flush = () => {
    if (!header) return;
    const cols = new Set(header);
    if (cols.has("role") && cols.has("direction")) {
      for (const data of rows) {
        channels.push({
          id: data.id ?? "",
          name: data.name ?? "",
          role: (data.role ?? "") as Role,
          direction: (data.direction ?? "") as Direction,
        });
      }
    } else if (cols.has("room") && cols.has("state")) {
      for (const data of rows) {
        work.push({
          id: data.id ?? "",
          name: data.name ?? "",
          room: data.room ?? "",
          state: data.state ?? "",
          anchor: data.anchor ?? "",
          owner: data.owner ?? "",
          pageId: data.page_id ?? "",
        });
      }
    }
    header = null;
    rows = [];
  };

  for (const raw of body.split("\n")) {
    const line = raw.trim();
    if (!line.startsWith("|")) {
      flush();
      continue;
    }
    const cells = line.replace(/^\||\|$/g, "").split("|").map((c) => c.trim());
    if (/^[-:\s|]+$/.test(line)) continue;
    if (!header) {
      header = cells.map((c) => c.toLowerCase());
      continue;
    }
    const data: Record<string, string> = {};
    header.forEach((key, i) => {
      data[key] = cells[i] ?? "";
    });
    rows.push(data);
  }
  flush();
  return { channels, work };
}

export function plan(mapping: Mapping): { work: Work; action: "post" | "gather" | "skip" | "gap" }[] {
  const rooms = new Set(mapping.channels.filter((c) => c.role === "room").map((c) => c.id));
  return mapping.work.map((w) => {
    if (!w.room || !rooms.has(w.room)) return { work: w, action: "gap" as const };
    if (w.state === "closed") return { work: w, action: "skip" as const };
    if (w.anchor) return { work: w, action: "gather" as const };
    return { work: w, action: "post" as const };
  });
}

export function replaceWork(mapping: Mapping, work: Work[]): Mapping {
  return { ...mapping, work };
}

function validate(meta: Record<string, string>, channels: Channel[], work: Work[]) {
  for (const key of ["workspace", "platform", "scope", "agent"]) {
    if (!meta[key]) throw new MappingError(`missing ${key}`);
  }
  if (meta.platform !== "slack" && meta.platform !== "discord") {
    throw new MappingError(`unknown platform ${meta.platform}`);
  }
  const ids = channels.map((c) => c.id);
  if (new Set(ids).size !== ids.length) throw new MappingError("duplicate channel id");
  const hasMgmt = channels.some((c) => c.role === "management" && c.direction !== "out");
  const hasReport = channels.some((c) => c.role === "reports" && c.direction !== "in");
  if (!hasMgmt) throw new MappingError("need a management channel with direction in/both");
  if (!hasReport) throw new MappingError("need a reports channel with direction out/both");
  const workIds = work.map((w) => w.id);
  if (new Set(workIds).size !== workIds.length) throw new MappingError("duplicate work id");
}

File: agent/python/mapping.py
"""Parse an Axiom mapping markdown file (front matter + table)."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from urllib.request import Request, urlopen

ROLES = {"management", "room", "reports"}
DIRECTIONS = {"in", "out", "both"}
PLATFORMS = {"slack", "discord"}
WORK_STATES = {"active", "blocked", "closed"}


@dataclass(frozen=True)
class Channel:
    id: str
    name: str
    role: str
    direction: str


@dataclass(frozen=True)
class Work:
    id: str
    name: str
    room: str
    state: str
    anchor: str
    owner: str
    page_id: str = ""


@dataclass(frozen=True)
class Mapping:
    workspace: str
    platform: str
    scope: str
    agent: str
    collect_seconds: int
    channels: tuple[Channel, ...]
    work: tuple[Work, ...]

    def ids(self) -> set[str]:
        return {c.id for c in self.channels}

    def by_role(self, role: str) -> tuple[Channel, ...]:
        return tuple(c for c in self.channels if c.role == role)

    def room_ids(self) -> set[str]:
        return {c.id for c in self.channels if c.role == "room"}

    def plan(self) -> tuple[tuple[Work, str], ...]:
        """Each work row is a broadcast. action is post, gather, skip, or gap."""
        rooms = self.room_ids()
        out: list[tuple[Work, str]] = []
        for w in self.work:
            if not w.room or w.room not in rooms:
                out.append((w, "gap"))
            elif w.state == "closed":
                out.append((w, "skip"))
            elif w.anchor:
                out.append((w, "gather"))
            else:
                out.append((w, "post"))
        return tuple(out)

    def replace_work(self, work: tuple[Work, ...]) -> Mapping:
        return Mapping(
            workspace=self.workspace,
            platform=self.platform,
            scope=self.scope,
            agent=self.agent,
            collect_seconds=self.collect_seconds,
            channels=self.channels,
            work=work,
        )


class MappingError(ValueError):
    pass


def read_source(source: str) -> str:
    if source.startswith(("http://", "https://")):
        req = Request(source, headers={"User-Agent": "axiom-relay/0.3"})
        with urlopen(req, timeout=15) as resp:
            return resp.read().decode("utf-8")
    return Path(source).read_text(encoding="utf-8")


def load_mapping_from(source: str) -> Mapping:
    return load_mapping(read_source(source))


def load_mapping(text: str) -> Mapping:
    meta, table = _split_front_matter(text)
    channels, work = _parse_tables(table)
    _validate(meta, channels, work)
    return Mapping(
        workspace=meta["workspace"],
        platform=meta["platform"],
        scope=meta["scope"],
        agent=meta["agent"],
        collect_seconds=int(meta.get("collect_seconds", 45)),
        channels=channels,
        work=work,
    )


def _split_front_matter(text: str) -> tuple[dict[str, str], str]:
    if not text.startswith("---"):
        raise MappingError("mapping must start with YAML front matter")
    rest = text[3:]
    end = rest.find("\n---")
    if end < 0:
        raise MappingError("unterminated front matter")
    raw, body = rest[:end], rest[end + 4 :]
    meta: dict[str, str] = {}
    for line in raw.splitlines():
        if not line.strip() or ":" not in line:
            continue
        key, value = line.split(":", 1)
        meta[key.strip()] = value.strip()
    return meta, body


def _parse_tables(body: str) -> tuple[tuple[Channel, ...], tuple[Work, ...]]:
    channels: list[Channel] = []
    work: list[Work] = []
    header: list[str] | None = None
    rows: list[dict[str, str]] = []

    def flush() -> None:
        nonlocal header, rows
        if header:
            cols = set(header)
            if "role" in cols and "direction" in cols:
                for data in rows:
                    channels.append(
                        Channel(
                            id=data.get("id", ""),
                            name=data.get("name", ""),
                            role=data.get("role", ""),
                            direction=data.get("direction", ""),
                        )
                    )
            elif "room" in cols and "state" in cols:
                for data in rows:
                    work.append(
                        Work(
                            id=data.get("id", ""),
                            name=data.get("name", ""),
                            room=data.get("room", ""),
                            state=data.get("state", ""),
                            anchor=data.get("anchor", ""),
                            owner=data.get("owner", ""),
                            page_id=data.get("page_id", ""),
                        )
                    )
        header = None
        rows = []

    for raw in body.splitlines():
        line = raw.strip()
        if not line.startswith("|"):
            flush()
            continue
        cells = [c.strip() for c in line.strip("|").split("|")]
        if set("".join(cells)) <= set("-: "):
            continue
        if header is None:
            header = [c.lower() for c in cells]
            continue
        rows.append(dict(zip(header, cells)))
    flush()
    return tuple(channels), tuple(work)


def _validate(
    meta: dict[str, str],
    channels: tuple[Channel, ...],
    work: tuple[Work, ...],
) -> None:
    for key in ("workspace", "platform", "scope", "agent"):
        if not meta.get(key):
            raise MappingError(f"missing {key}")
    if meta["platform"] not in PLATFORMS:
        raise MappingError(f"unknown platform {meta['platform']!r}")
    ids = [c.id for c in channels]
    if len(ids) != len(set(ids)):
        raise MappingError("duplicate channel id")
    for ch in channels:
        if ch.role not in ROLES:
            raise MappingError(f"unknown role {ch.role!r}")
        if ch.direction not in DIRECTIONS:
            raise MappingError(f"unknown direction {ch.direction!r}")
    if not any(c.role == "management" and c.direction in {"in", "both"} for c in channels):
        raise MappingError("need a management channel with direction in/both")
    if not any(c.role == "reports" and c.direction in {"out", "both"} for c in channels):
        raise MappingError("need a reports channel with direction out/both")
    work_ids = [w.id for w in work]
    if len(work_ids) != len(set(work_ids)):
        raise MappingError("duplicate work id")
    for w in work:
        if w.state and w.state not in WORK_STATES:
            raise MappingError(f"unknown work state {w.state!r}")
